src/Facility.Core/Http/HttpClientService.cs

```diff
@@ -101,8 +103,66 @@ protected HttpClientService(HttpClientServiceSettings? settings, Uri? defaultBas
 			{
 				var contentType = mapping.RequestBodyContentType ?? requestHeaders?.GetContentType();
 				httpRequest.Content = GetHttpContentSerializer(requestBody.GetType()).CreateHttpContent(requestBody, contentType);
-				if (m_disableChunkedTransfer)
+				if (m_shouldCompressRequest(request))
+					httpRequest.Content = await CompressContentAsync(httpRequest.Content, cancellationToken).ConfigureAwait(false);
+				else if (m_disableChunkedTransfer)
 					await httpRequest.Content.LoadIntoBufferAsync().ConfigureAwait(false);
+
+				async static Task<HttpContent> CompressContentAsync(HttpContent httpContent, CancellationToken cancellationToken)
+				{
+					Stream? contentStream = null;
+					Stream? compressedContentStream = null;
+
+					try
+					{
+						// copy the existing HTTP content into a memory stream
+						contentStream = new MemoryStream();
+#if NET5_0_OR_GREATER
+						await httpContent.CopyToAsync(contentStream, cancellationToken).ConfigureAwait(false);
```

Comment:
Could leverage `LoadIntoBufferAsync` instead?

Reply:
Do you mean `CreateContentReadStreamAsync`? I'm not sure how calling `LoadIntoBufferAsync` provides access to the data.

Reply:
Oh, that's `protected`. I guess `LoadIntoBufferAsync` then `ReadAsStreamAsync`?

Reply:
`LoadIntoBufferAsync` reads the content into an array-pooled internal memory stream, so you could read it twice. It should be the same thing that you are doing, except perhaps more efficient.

Reply:
`LoadIntoBufferAsync` may be a premature optimization? Just use `ReadAsStreamAsync` and let it pick the optimal implementation?

Reply:
No, wait, I still want to have a `MemoryStream` for the case where we decide that compression isn't worth it and want to send the uncompressed data. I need to be able to rewind the Stream and use it for HttpContent, which might not be possible with an arbitrary Stream.

Reply:
Presumably you can just open it again instead of rewinding it.

I'm not sure that deciding that compression isn't worth it is worth all of that memory usage.

If we're worried about small payloads, we could always have special logic for requests that fit in the first 64K block (or whatever we use).